File: crates/varve-effects/src/palette_snap.rs
//! Palette Snap kernel — faithful port of
//! `packages/engine/src/liveEffects/paletteSnap.ts`.
//!
//! Nearest-colour matching can run in RGB, linear RGB, Lab, or OKLab, with an
//! optional dither pass (reusing the Dither kernel's primitives). Transparent
//! pixels stay transparent unless the palette itself has no alpha concept.

use crate::palette_core::{self, ColorMetric, PaletteLookup};
use crate::{clamp01, clamp_byte, CoordSpace, Params};

/// Apply palette snapping in place — port of `applyPaletteSnap`.
pub fn apply(out: &mut [u8], w: u32, h: u32, p: &Params) {
    if w == 0 || h == 0 {
        return;
    }

    let raw = p.rgb_list("colors");
    let colors_f64: Vec<[f64; 3]> =
        raw.iter().map(|c| [c[0] as f64, c[1] as f64, c[2] as f64]).collect();
    let colors = palette_core::sanitize_palette(&colors_f64);
    let amount = clamp01(p.f("amount", 1.0));
    let alpha_cutoff = clamp01(p.f("alphaCutoff", 0.0));
    if colors.is_empty() || amount <= 0.0 {
        return;
    }

    let metric = ColorMetric::parse(p.s("metric", "rgb"));
    let lookup = PaletteLookup::build(&colors, metric);

    let len = (w * h) as usize;
    for i in 0..len {
        let o = i * 4;
        let a = out[o + 3];
        if (a as f64) / 255.0 < alpha_cutoff || a == 0 {
            continue;
        }
        let c = lookup.find(out[o], out[o + 1], out[o + 2]);
        if amount >= 1.0 {
            out[o] = c[0];
            out[o + 1] = c[1];
            out[o + 2] = c[2];
        } else {
            out[o] = clamp_byte(out[o] as f64 + (c[0] as f64 - out[o] as f64) * amount);
            out[o + 1] =
                clamp_byte(out[o + 1] as f64 + (c[1] as f64 - out[o + 1] as f64) * amount);
            out[o + 2] =
                clamp_byte(out[o + 2] as f64 + (c[2] as f64 - out[o + 2] as f64) * amount);
        }
    }

    if p.b("dither", false) && p.f("ditherStrength", 0.0) > 0.0 {
        let deduped = palette_core::dedupe_palette(&colors);
        let dither_params = serde_json::json!({
            "algorithm": p.s("ditherAlgorithm", "floyd-steinberg"),
            "paletteMode": "custom",
            "levels": 4,
            "colors": deduped,
            "metric": p.s("metric", "rgb"),
            "serpentine": true,
            "strength": p.f("ditherStrength", 0.0),
            "bayerSize": 4,
            "cellSize": 1,
            "alphaCutoff": alpha_cutoff,
            "seed": p.f("seed", 0.0),
        });
        let dp = Params::new(&dither_params);
        // TS passes its `coordSpace` through, which paletteSnap callers leave
        // absent — the default (all-zero) CoordSpace anchors to the surface
        // grid exactly like `undefined` does in JS.
        crate::dither::apply(out, w, h, &dp, CoordSpace::default());
    }
}

File: packages/engine/src/liveEffects/paletteSnap.ts
/**
 * Palette Snap — non-destructive quantization to an explicit palette.
 *
 * Never changes the document's colour mode. The palette is stored inline on
 * the effect (copied from document swatches or an imported .gpl/.ase/.aco/.act
 * file at edit time), so rendering is fully deterministic and standalone.
 *
 * Nearest-colour matching can run in RGB, linear RGB, Lab, or OKLab, with an
 * optional dither pass (reusing the Dither kernel's primitives). Transparent
 * pixels stay transparent unless the palette itself has no alpha concept.
 */

import { applyDither, type CoordSpace, type DitherAlgorithm } from './dither';
import type { ColorMetric } from './paletteCore';
import { buildPaletteLookup, dedupePalette, sanitizePalette } from './paletteCore';

export interface PaletteSnapParams {
  /** Palette colours as [r, g, b]. */
  colors: readonly (readonly number[])[];
  /** Matching metric. */
  metric: ColorMetric;
  /** Mix the snapped colour with the source (0 = unchanged, 1 = full snap). */
  amount: number;
  /** Whether to dither the quantization error. */
  dither: boolean;
  ditherAlgorithm: DitherAlgorithm;
  /** 0..1 dither strength. */
  ditherStrength: number;
  /** Pixels below this alpha are forced fully transparent. */
  alphaCutoff: number;
  seed: number;
}

/** Apply palette snapping in place. Returns the same ImageData. */
export function applyPaletteSnap(
  imageData: ImageData,
  params: PaletteSnapParams,
  coordSpace?: CoordSpace,
): ImageData {
  const { data, width: w, height: h } = imageData;
  if (w === 0 || h === 0) return imageData;

  const colors = sanitizePalette(params.colors ?? []);
  const amount = clamp01(params.amount ?? 1);
  const alphaCutoff = clamp01(params.alphaCutoff ?? 0);
  if (colors.length === 0 || amount <= 0) return imageData;

  const lookup = buildPaletteLookup(colors, params.metric ?? 'rgb');

  for (let i = 0; i < data.length; i += 4) {
    const a = data[i + 3]!;
    if (a / 255 < alphaCutoff || a === 0) continue;
    const c = lookup.find(data[i]!, data[i + 1]!, data[i + 2]!);
    if (amount >= 1) {
      data[i] = c[0];
      data[i + 1] = c[1];
      data[i + 2] = c[2];
    } else {
      data[i] = Math.round(data[i]! + (c[0] - data[i]!) * amount);
      data[i + 1] = Math.round(data[i + 1]! + (c[1] - data[i + 1]!) * amount);
      data[i + 2] = Math.round(data[i + 2]! + (c[2] - data[i + 2]!) * amount);
    }
  }

  if (params.dither && params.ditherStrength > 0) {
    applyDither(
      imageData,
      {
        algorithm: params.ditherAlgorithm ?? 'floyd-steinberg',
        paletteMode: 'custom',
        levels: 4,
        colors: dedupePalette(colors),
        metric: params.metric ?? 'rgb',
        serpentine: true,
        strength: params.ditherStrength,
        bayerSize: 4,
        cellSize: 1,
        alphaCutoff,
        seed: params.seed ?? 0,
      },
      coordSpace,
    );
  }
  return imageData;
}

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}
